Add JoinTestRunner to thoroughly test join operators

mrzzzrm opened the pull request:
This is to fix #1542, it runs each of our JoinOperators (NestedLoop, Hash, SortMerge, MPSM, Index-without-indices-for-now) in up to ~1000 configurations (depending on the operations supported by the Operator).

Also, fix #1557

### Bugs hunted down so far:
* JoinNestedLoop had problems with Right/Full Outer Joins and an empty left table
* JoinIndex didn't support Semi/Anti in its fallback-to-jnl mode
* JoinSortMerge couldn't deal with empty input table(s) Part I: `ClusterCount must be > 0`
* JoinSortMerge couldn't deal with empty input table(s) Part II: `Every partition is empty`
* JoinSortMerge sometimes emitted tuples filled up with NULL multiple times.
* JoinHash was buggy for an obscure AntiNullAsTrue cases
* JoinSortMerge/MPSM had UB when columns contained values > UINT_MAX, since `get_radix` would cast every arithmetic value to `uint32_t`

mrks commented:
> Ship result tables in git repo or generate them on the fly as the tests are running? If we ship the tables that's a lot storage required (1.1 MB) and, more importantly, the generator scripts would not be run in CI - we don't want code in our repo that's not run in CI.

How would these tables be different from the test tables that we already have? It's not that these will change any time soon. [Github considers](https://help.github.com/en/articles/working-with-large-files) files with more than 50 MB as big and with more than 100 MB as too big. I'm slightly leaning towards simply committing them, even if this means that you will save your most-lines-added-spot in the contributor list for years to come. ;)

> Generate input/output tables as well as configurations with python scripts or with c++? The latter would be cleaner to invoke from c++.

This largely depends on the answer to the previous question. If they need to be generated as part of the test, it should be C++. Otherwise, I would think that its easier to write in Python. If there is a python-supported database that supports all join types (i.e., not sqlite), we could even cross-check our generation there.

> If we generate from c++, do we generate tables and configurations as files? I find this important because having actual tbl files to look at is a great thing when debugging failing tests. Having the configurations as a file is great for validating that a case is tested.

Another reason to generate this outside of the tests.

> On my 4Ghz machine the tests take around 3s, significantly more on my laptop. Do we want to burden this on the unit test suite, or do we run the JoinTestRunner as part of the system test?

I can live with either. Maybe we can execute some configurations (e.g., segments other than ValueSegments) only in the system test?

mrks commented:
If the coverage provided by these tests is good enough, #1409 should be fixed, correct?

mrzzzrm commented:
Nope, only partially fixed in the best case. The main point of #1409 is "Some (e.g. 10a) do not finish within reasonable time (left running 3h) [in sqlite]"

mrks commented:
See #1617.

mrzzzrm commented:
Now - given the info that a PR fixing the other half of #1409 exists - the answer to your question would be: Yes ;)

mrzzzrm commented:
@mrks ready for review. Better name for `JoinReferenceOperator` is still pending, I haven't come up with anything yet. Paging @Bouncner to notify you that all join operators changed at least a bit.

mrzzzrm commented on src/lib/operators/join_reference_operator.cpp:
```diff
@@ -0,0 +1,220 @@
+#include "join_reference_operator.hpp"
```
(random line) @Bouncner @mrks What do you think of `JoinVerificationOperator` as a name? The term "verify" is used with similar semantic in the `--verify` option of the benchmarks, that's why I think it may be a good fit for something that provides a ground truth.

Bouncner replied:
Sounds good to me. I like the term especially since I don’t think somebody will use this operator accidentally. It’s clearer than reference.

mrzzzrm replied:
`JoinVerification` without "*Operator" suffix it is. Suffix dropped due to @mrks implicit dismissal https://github.com/hyrise/hyrise/pull/1577#discussion_r279195268

Bouncner commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,530 @@
+    // The number of radix bits affects the partitioning inside the JoinHash operator, thus table/chunk sizes and
+    // different join modes need to be tested in combination, since they all use this partitioning.
+    if constexpr (std::is_same_v<JoinOperator, JoinHash>) {
+      for (const auto radix_bits : {1, 2, 5}) {
```
Is `radix_bits = 0` tested as well somewhere?

mrzzzrm replied:
I probably assumed that was the default, but it is not. I added it here, let's see whether there are bugs ;)

mrzzzrm replied:
all seems fine

Bouncner commented on src/benchmarklib/benchmark_runner.cpp:
```diff
@@ -371,10 +371,12 @@ void BenchmarkRunner::_execute_query(const QueryID query_id, const std::shared_p
+                                       FloatComparisonMode::RelativeDifference)) {
         _query_results[query_id].verification_passed = false;
-        std::cout << "- Verification failed (" << timer.lap_formatted() << ")" << std::endl;
+        std::cout << "- Verification failed (" << timer.lap_formatted() << ")" << std::endl
```
Any reason to use `std::endl` in between over `\n`?

mrzzzrm replied:
Na, just habit. Wouldn't change it here to avoid conflicts with #1597

Bouncner commented on src/lib/operators/join_hash.cpp:
```diff
@@ -21,54 +21,69 @@
+   * JoinMode::Inner        The smaller relation becomes the build side, the bigger the probe side
+   * JoinMode::Left/Right   The outer relation becomes the probe side, the inner relation becomes the build side
+   * JoinMode::FullOuter    Not supported by JoinHash
+   * JoinMode::Semi/Anti*   The inner relation becomes the build side, the outer relation becomes the probe side
```
Is there an inner and outer relation for semi/anti joins?

mrzzzrm replied:
Mmh, they are not called like that, right? reverting to left/right then.

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -211,44 +211,45 @@ std::vector<std::optional<HashTable<HashedType>>> build(const RadixContainer<Lef
-      auto& partition_left = static_cast<Partition<LeftType>&>(*radix_container.elements);
+    jobs.emplace_back(std::make_shared<JobTask>(
+        [&, build_partition_begin, build_partition_end, current_partition_id, build_partition_size]() {
+          auto& build_partition = static_cast<Partition<BuildColumnType>&>(*radix_container.elements);
```
Is that indentation off?

mrzzzrm replied:
Na, it's okay. I wouldn't clang-format-off the entire Job-lambda just to get the lambda header right

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -533,57 +533,80 @@ void probe_semi_anti(const RadixContainer<RightType>& radix_container,
-          auto& row = partition[partition_offset];
+          auto& probe_column_element = partition[partition_offset];
 
-          if constexpr (retain_null_values) {
```
A particular reason to remove the template parameter?

mrzzzrm replied:
It just wasn't needed anymore.

mrzzzrm replied:
I replaced it with a JoinMode template param now.

Bouncner commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -596,20 +619,20 @@ void probe_semi_anti(const RadixContainer<RightType>& radix_container,
  */
 // See usage in _on_execute() for doc.
-inline PosListsBySegment setup_pos_lists_by_segment(const std::shared_ptr<const Table>& input_table) {
+inline PosListsByChunk setup_pos_lists_by_segment(const std::shared_ptr<const Table>& input_table) {
```
Return type and function name are confusing. Is it now by chunk or segment?

mrzzzrm replied:
Yep, changed return type name without changing the function name. Fixed

Bouncner commented on src/lib/operators/join_sort_merge.cpp:
```diff
@@ -39,19 +46,13 @@ JoinSortMerge::JoinSortMerge(const std::shared_ptr<const AbstractOperator>& left
-  Assert(mode != JoinMode::Cross, "This operator does not support cross joins.");
-  Assert(left, "The left input operator is null.");
-  Assert(right, "The right input operator is null.");
   Assert(primary_predicate.predicate_condition == PredicateCondition::Equals ||
```
Your reason to keep that assert?

mrzzzrm replied:
I moved it to AbstractJoinOperator's constructor. We should have some place making sure JoinOps are not called with IN/EXISTS predicates

Bouncner commented on src/lib/operators/join_sort_merge.cpp:
```diff
@@ -182,7 +189,8 @@ class JoinSortMerge::JoinSortMergeImpl : public AbstractJoinOperatorImpl {
     // Note: this is only provisional. There should be a reasonable calculation here based on hardware stats.
     size_t chunk_count_left = _sort_merge_join.input_table_left()->chunk_count();
     size_t chunk_count_right = _sort_merge_join.input_table_right()->chunk_count();
-    return static_cast<size_t>(std::pow(2, std::floor(std::log2(std::max(chunk_count_left, chunk_count_right)))));
```
Reason for that change? Since the chunk counts can never be negative, there shouldn't be values smaller than 2^0. One problem that might occur are empty tables with chunk counts of 0 such that `log2()` fails, but that's still not covered (can that even happen?).

mrzzzrm replied:
I think I changed this precisely because Tables might be empty, i.e., containing no Chunks. Yes, imo (and I *think* @mrks agrees) this is a very valid state and operators need to handle it properly

Bouncner replied:
Not what I meant. I see the point for empty chunks, no question.

But what the current change does it to ensure that at least a value of 1 is returned. However, 2^0 is 1, so that's ensured anyhow. Of course assuming we will never see negative chunk counts (should be save to assume). Another problem that is currently not handled (and hasn't been before) is the `log(0)` case, which would fail (both tables have 0 chunks). So what I am proposing is to move the` max(1)` to the chunk count max: `std::log2(std::max({1, chunk_count_left, chunk_count_right})`. For all values >= 1, the following `2^n` will always yield a valid chunk count.

mrzzzrm replied:
gottcha. changed.

Bouncner commented on src/lib/operators/join_sort_merge/radix_cluster_sort.hpp:
```diff
@@ -66,6 +68,26 @@ class RadixClusterSort {
+  static std::enable_if_t<std::is_same_v<T2, pmr_string>, uint32_t> get_radix(T2 value, size_t radix_bitmask) {
+    PerformanceWarning("Using memcpy to perform bit_cast/radix partitioning of strings");
+    uint32_t radix;
+    std::memcpy(&radix, value.c_str(), std::min(value.size(), sizeof(radix)));
```
Reason to not use std::hash here?

mrzzzrm replied:
strings used memcpy before. But this is actually UB, as the memcpy might access memory it shouldn't for strings <= 3 chars. will use std::hash then.

Bouncner commented on src/lib/operators/print.hpp:
```diff
@@ -8,7 +8,8 @@
 namespace opossum {
-enum PrintFlags { PrintMvcc = 1 << 1 };
+
+enum PrintFlags { PrintMvcc = 1 << 1, PrintIgnoreChunkBoundaries = 1u << 2u };
```
Haven't heard of these chunk boundaries before. It's just about whether we print where a chunk ends and the other begins?

mrzzzrm replied:
precisely. For chunk_size = 1 (which happens in tests) it's hard to decipher table printings without this flag. will doc it.

Bouncner commented on src/lib/storage/table.cpp:
```diff
@@ -193,6 +194,51 @@ void Table::append_chunk(const std::shared_ptr<Chunk>& chunk) {
 }
 
+std::vector<AllTypeVariant> Table::get_row(size_t row_idx) const {
+  for (const auto& chunk : _chunks) {
```
Performance warnings for both new methods? They look by far too nice to be not used by students.

mrzzzrm replied:
yesss

Bouncner commented on src/lib/operators/join_sort_merge/radix_cluster_sort.hpp:
```diff
@@ -66,6 +68,17 @@ class RadixClusterSort {
+  }
+
+  template <typename T2>
+  static std::enable_if_t<!std::is_integral_v<T2>, size_t> get_radix(T2 value, size_t radix_bitmask) {
```
`is_floating_point`?

Bouncner replied:
And function for `pmr_string` appears to be missing?

mrzzzrm replied:
since float and string now both use std::hash, there is no need anymore for dedicated pmr_string/is_floating_point impls. they just share this one

Bouncner replied:
🤦

Bouncner commented on src/lib/operators/operator_join_predicate.cpp:
```diff
@@ -53,4 +53,8 @@ void OperatorJoinPredicate::flip() {
 }
 
+bool operator<(const OperatorJoinPredicate& l, const OperatorJoinPredicate& r) {
+  return std::make_tuple(l.column_ids, l.predicate_condition) < std::make_tuple(r.column_ids, r.predicate_condition);
```
Is the creation of a tuple really necessary here?

mrzzzrm replied:
less writing. It most probably won't be "created" at runtime

Bouncner replied:
I'd guess it's seldomly used? And just within the optimizer? 

If not, I would rather make sure that we don't create tuple unnecessarily.

mrzzzrm replied:
I'd bet a good chunk of... let's say a finger of my left hand that no extra machine-code is generated for this in release builds. And even if it were, this is not used in hot loops.

mrks commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,527 @@
+  InputTableConfiguration input_left;
+  InputTableConfiguration input_right;
+  JoinMode join_mode{JoinMode::Inner};
+  DataType data_type_left{DataType::Int};
```
These defaults seem somewhat arbitrary. Would it be clearer if those parameters were explicitly set?

mrzzzrm replied:
When varying only a subset of the dimensions of the configuration, there have to be *some* defaults, we just don't get around that. Explicitly typing all defaults of in each place where a configuration is generated, is not better: It generates noise and makes it look like all those values were somehow chosen deliberately when they were actually not:

Compare this
```c++
    // Test all combinations of reference/data input tables. This tests mostly the composition of the output table
    for (const auto& left_input_table_type : all_input_table_types) {
      for (const auto& right_input_table_type : all_input_table_types) {
        auto join_test_configuration = default_configuration;
        join_test_configuration.input_left.table_type = left_input_table_type;
        join_test_configuration.input_right.table_type = right_input_table_type;

        add_configuration_if_supported(join_test_configuration);
      }
    }
```

to this
```c++
 // Test all combinations of reference/data input tables. This tests mostly the composition of the output table
for (const auto& left_input_table_type : all_input_table_types) {
  for (const auto& right_input_table_type : all_input_table_types) {
    auto join_test_configuration = JoinTestConfiguration;
    join_test_configuration.input_left.input_side = InputSide::Left;
    join_test_configuration.input_left.chunk_size = all_chunk_sizes.front();
    join_test_configuration.input_left.table_size = 10;
    join_test_configuration.input_left.table_type = left_input_table_type;
    join_test_configuration.input_left.input_side = InputSide::Right;
    join_test_configuration.input_left.chunk_size = all_chunk_sizes.front();
    join_test_configuration.input_left.table_size = 10;
    join_test_configuration.input_right.table_type = right_input_table_type;
    join_test_configuration.input_right.join_mode = JoinMode::Inner;
    join_test_configuration.input_right.data_type_left = DataType::Int;
    join_test_configuration.input_right.data_type_right = DataType::Int;
    join_test_configuration.input_right.nullable_left = true;
    join_test_configuration.input_right.nullable_right = true;
    join_test_configuration.predicate_condition = PredicateCondition::Equals;
    join_test_configuration.secondary_predicates = all_secondary_predicate_sets.front();
    join_test_configuration.operator_factory = std::make_shared<JoinOperatorFactory<JoinOperator>>();

    add_configuration_if_supported(join_test_configuration);
  }
}

I want it to be discernable which configuration parameter was chosen deliberately and which was not.
```

mrks replied:
Ok

mrks commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,527 @@
+                                                        const JoinMode mode,
+                                                        const OperatorJoinPredicate& primary_predicate,
+                                                        const std::vector<OperatorJoinPredicate>& secondary_predicates,
+                                                        const JoinTestConfiguration& configuration) override {
```
Aren't all of the above part of the configuration?

mrzzzrm replied:
well, at least join_mode and seconary_predicates are direct members of the configuration. All other ones need to be built from info in the configuration.

mrks replied:
For me, it’s confusing to pass the same parameter twice. What keeps someone from using the `mode` from `configuration`, which might differ from the one passed in?

mrzzzrm replied:
sorry, I wasn't clear. Of course I removed `join mode` and `secondary predicates` from the parameter list and use the one from `configuration`

mrks commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,527 @@
+
+    /**
+     * For each `all_*` set, the first element is the default - which is chosen to be sensible. Thus the elements are
+     * not necessarily sorted.
```
What do "default" and "sorted mean in this context?

mrzzzrm replied:
Update comment to 
```
    /**
     * For each `all_*` set, the first element is used to build the default/base configuration 
     * (`JoinTestConfiguration default_configuration`). Thus the first elements are chosen to be sensible defaults.
     */
```

("sorted" meant just that: "Don't confused if it's not sorted, the elements are ordered such that the first is a good default")

mrks commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,527 @@
+      DataType::Int,
+      DataType::Int,
+      true,
+      true,
```
Enable c++20 and bring in designated initializers?

mrzzzrm replied:
Given how red/orange clang still is, I'd say no: https://clang.llvm.org/cxx_status.html

mrks commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,527 @@
+      }
+    }
+
+    // JoinHash specific configurations varying the number of radix_bits.
```
Does it make sense to test pre-sorted / unsorted for JoinSortMerge @Bouncner ?

mrzzzrm replied:
Unless there is a special "detect_sorted_input" path in the sorting algo, I'd say "no". Firing random configuration combinations into test runner makes its runtime explode pretty quickly.

mrks commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,527 @@
+      }
+    }
+
+    // Test MPJ support for all join modes. Swap the input tables to trigger cases especially in the Anti* modes
```
Abbrevation

mrzzzrm commented on resources/test_data/sqlite_testrunner_queries.sql:
```diff
@@ -136,8 +136,8 @@ SELECT * FROM mixed AS t1 LEFT JOIN mixed_null AS t2 ON t1.a = t2.a AND t1.b <=
 SELECT * FROM id_int_int_int_100 AS t1 LEFT JOIN id_int_int_int_100 AS t2 ON t1.a = t2.a;
--- #1527 SELECT * FROM id_int_int_int_100 AS t1 LEFT JOIN id_int_int_int_100 AS t2 ON t1.a > t2.a AND t1.b >= t2.b AND t1.c < t2.c;
--- #1527 SELECT * FROM id_int_int_int_100 AS t1 LEFT JOIN id_int_int_int_100 AS t2 ON t1.a <= t2.a AND t1.b > t2.b AND t1.c < t2.c;
+SELECT * FROM id_int_int_int_100 AS t1 LEFT JOIN id_int_int_int_100 AS t2 ON t1.a > t2.a AND t1.b >= t2.b AND t1.c < t2.c;
```
Doesn't even have to be bigger input tables, imo. And yes, should be possible. But I'd not do it as part of this PR, because I don't want to fix #1643 as part of this PR and because I'd like to keep my pipeline small. Opened issue #1659

mrzzzrm commented on src/lib/operators/join_hash.cpp:
```diff
@@ -21,54 +21,76 @@
+const std::string JoinHash::description(DescriptionMode description_mode) const {
+  std::ostringstream stream;
+  stream << AbstractJoinOperator::description(description_mode);
+  stream << " RadixBits: " << (_radix_bits ? std::to_string(*_radix_bits) : "Unset");
```
No particular reason. Changed to "Radix bits"

mrzzzrm commented on src/lib/operators/join_hash.cpp:
```diff
@@ -21,54 +21,76 @@
     // We don't have to swap the operation itself here, because we only support the commutative Equi Join.
-    build_operator = _input_right;
-    probe_operator = _input_left;
+    build_input_operator = _input_right;
```
was there before, but I'll refactor it

mrzzzrm commented on src/lib/operators/join_hash.cpp:
```diff
@@ -113,21 +136,21 @@ std::shared_ptr<const Table> JoinHash::_on_execute() {
 void JoinHash::_on_cleanup() { _impl.reset(); }
 
-template <typename LeftType, typename RightType>
+template <typename BuildColumnType, typename ProbeColumnType>
```
yes, thoroughly I'd claim

mrzzzrm commented on src/lib/operators/join_hash.cpp:
```diff
@@ -138,19 +161,20 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
   const ColumnIDPair _column_ids;
   const PredicateCondition _predicate_condition;
-  const bool _inputs_swapped;
+  // If true, the output table will have the probe side columns left of the build side columns.
```
Changed it to an enum class to correctly image the ternary nature of which input table gets output where:

```c++
// Depending on which input table became the build/probe table we have to order the columns of the output table.
// Semi/Anti* Joins only emit tuples from the probe table
enum class OutputColumnOrder {
  BuildFirstProbeSecond,
  ProbeFirstBuildSecond,
  ProbeOnly
};
```

mrzzzrm commented on src/lib/operators/join_hash.cpp:
```diff
@@ -258,133 +292,154 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
-        radix_left = partition_radix_parallel<LeftType, HashedType, false>(materialized_left, left_chunk_offsets,
-                                                                           histograms_left, _radix_bits);
+        // radix partition the build table
+        if (keep_nulls_build_column) {
```
wholeheartedly agree. I want `std::resolve(<enum_or_bool>, Functor functor)`

mrzzzrm commented on src/lib/operators/join_hash.cpp:
```diff
@@ -200,49 +224,59 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
-     * In the current implementation, the relation on the right is always the outer relation.
-     * The AntiNullAsFalse-Join, too, will emit tuples with a NULL
+    /**
+     * Discard NULLs from build and probe columns as follows
```
Changed to "Keep/Discard NULLs from build and probe columns as follows"

mrzzzrm commented on src/lib/operators/join_hash.cpp:
```diff
@@ -258,133 +292,154 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
+    // (Hacky) Short cut for AntiNullAsTrue
+    //          If there is any NULL value on the build side, do not bother probe as no tuples can be emitted
+    //          anyway (as long as JoinHash/AntiNullAsTrue doesn't support secondary predicates). Doing this early out
+    //          here is hacky, but during probing we assume NULL values on the build
```
Changed as suggested

mrzzzrm commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -189,19 +189,19 @@ RadixContainer<T> materialize_input(const std::shared_ptr<const Table>& in_table
   /*
   NUMA notes:
-  The hashtables for each partition P should also reside on the same node as the two vectors leftP and rightP.
+  The hashtables for each partition P should also reside on the same node as the two vectors buildP and probeP.
```
As far as I can tell they never did, I took these "NUMA notes" comments as "look out for this if you ever make this operator NUMA friendly". @lanice.

mrzzzrm commented on src/lib/operators/join_hash/join_hash_steps.hpp:
```diff
@@ -533,57 +533,80 @@ void probe_semi_anti(const RadixContainer<RightType>& radix_container,
+          pos_list_local.emplace_back(probe_column_element.row_id);
+        }
+      } else if constexpr (mode == JoinMode::AntiNullAsTrue) {  // NOLINT - doesn't like else if constexpr
         // no hashtable on other side, but we are in Anti mode which means all tuples from the probing side get emitted
```
The subtle difference is that for JoinAntiNullAsTrue: `... all tuples from the probing side get emitted. That is, except NULL values, which only get emitted if the build table is empty.`

mrzzzrm commented on src/lib/operators/join_index.cpp:
```diff
@@ -56,18 +66,23 @@ void JoinIndex::_perform_join() {
   _left_matches.resize(input_table_left()->chunk_count());
 
-  const auto track_left_matches = (_mode == JoinMode::Left || _mode == JoinMode::FullOuter);
+  const auto is_outer_join = _mode == JoinMode::Left || _mode == JoinMode::Right || _mode == JoinMode::FullOuter;
```
Only used by JNL/JI and JI is under refactoring/revival anyway. Wouldn't change/refactor for something so minor. One could even claim they are the same anyway.

mrzzzrm commented on src/lib/operators/join_index.cpp:
```diff
@@ -231,39 +262,6 @@ void JoinIndex::_join_two_segments_using_index(LeftIterator left_it, LeftIterato
   }
 }
 
-// join loop that joins two segments of two columns via their iterators
```
apparently so.

mrzzzrm commented on src/lib/operators/join_mpsm.cpp:
```diff
@@ -36,19 +36,30 @@ STRONG_TYPEDEF(uint32_t, ClusterID);
+                   const OperatorJoinPredicate& primary_predicate,
+                   const std::vector<OperatorJoinPredicate>& secondary_predicates)
+    : AbstractJoinOperator(OperatorType::JoinMPSM, left, right, mode, primary_predicate, secondary_predicates) {
+  // TODO(moritz) incorporate into supports()?
```
Dear past self, it's creepy how you do that.

mrzzzrm commented on src/lib/operators/join_mpsm/column_materializer_numa.hpp:
```diff
@@ -16,25 +18,25 @@
-struct MaterializedValue {
-  MaterializedValue() = default;
-  MaterializedValue(RowID row, T v) : row_id{row}, value{v} {}
+struct MaterializedValueNUMA {
```
The *NUMA suffix was used before to differentiate JMPSM and JSM, e.g. ColumnMaterializationNUMA. Imo there is little code-cleanliness to gain here from minor naming things, JMPSM/JSM need a down-to-earth refactorwrite.

mrzzzrm commented on src/lib/storage/table.cpp:
```diff
@@ -193,6 +194,55 @@ void Table::append_chunk(const std::shared_ptr<Chunk>& chunk) {
   _chunks.push_back(chunk);
 }
 
+std::vector<AllTypeVariant> Table::get_row(size_t row_idx) const {
```
This is supposed to be used in nice-to-read code, so I'd keep the name simple. You'll get a performance warning, anyway.

mrzzzrm commented on src/lib/storage/table.cpp:
```diff
@@ -193,6 +194,55 @@ void Table::append_chunk(const std::shared_ptr<Chunk>& chunk) {
+  Fail("row_idx out of bounds");
+}
+
+std::vector<std::vector<AllTypeVariant>> Table::get_rows() const {
```
because I need it elsewhere, too, e.g., the JoinVerification[Operator].

mrzzzrm commented on src/lib/utils/check_table_equal.hpp:
```diff
@@ -11,8 +11,8 @@ namespace opossum {
- * "Strict" enforces that both tables have precisely the same column types, "Lenient" allows float instead of double, double
- * instead of float, long instead of int, int instead of long
+ * "Strict" enforces that both tables have precisely the same column types, "Lenient" allows float instead of double,
+ * double instead of float, long instead of int, int instead of long
```
Generally, I'd like for PR comments to only refer to things actually changed by that PR, not formatting changes etc. Otherwise people might get discouraged to work on existing unclean code because they fear they might be made responsible for the entire mess in their PR.

Anyway, added "We need this for comparing Hyrise with SQLite 
 * since the column types of the latter might differ from Hyrise's."

mrks replied:
I try not to place an undue burden upon whoever opens the PR when it comes to improving legacy code. At the same time, opening an issue for something this small seems like overkill, too.

I’ll keep your concern in mind and will both add “(or open an issue)” to more of the comments and try to limit the requests that actually take time.

mrzzzrm replied:
I'd just ignore minor things.

mrzzzrm commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,527 @@
+enum class InputTableType {
+  // Input Tables are data
+  Data,
+  // Input Tables are reference Tables with all Segments of a Chunk having the same PosList
```
Added 
```
// Join operators will build internal PosLists that they have to de-reference when assembling the output Table if the 
// input itself is already a reference Table.
```

mweisgut commented on src/lib/operators/join_hash.cpp:
```diff
@@ -21,54 +21,76 @@
 
 namespace opossum {
 
+bool JoinHash::supports(JoinMode join_mode, PredicateCondition predicate_condition, DataType left_data_type,
```
What's the crucial benefit with declaring the `supports`-function as static? Since all `AbstractJoinOperator`-derived classes implement this function, `supports(…)` could be implemented in `AbstractJoinOperator` as (non-static) pure virtual function to guarantee that possible upcoming derived classes have implemented this function.
Imo `supports` is in the same semantic level as `name()` of `AbstractOperator` but for join operators.

mrzzzrm replied:
nothings "crucial" ;)

The reasons I choose static over virtual are:
* With static there is no need to build a join operator that we never use; no need to build a join operator with a configuration that it doesn't support. Not (temporarily) constructing invalid state is what's probably most important to me here.
* Implementation of supports() is sufficiently (implicitly) enforced by places calling `JoinOperator::supports(...)` - so there is no realtistic way you could add a new join operator (which is added to the JoinTestRunner) without implementing supports()
* You'd save passing join_mode, secondary_predicates and predicate_condition with virtual, granted. But you'd not get around to pass left/right_data_type.

mweisgut commented on src/lib/operators/join_index.cpp:
```diff
@@ -56,18 +66,23 @@ void JoinIndex::_perform_join() {
+  const auto is_semi_or_anti_join =
+      _mode == JoinMode::Semi || _mode == JoinMode::AntiNullAsFalse || _mode == JoinMode::AntiNullAsTrue;
+
+  const auto track_left_matches = is_outer_join || is_semi_or_anti_join;
```
Tracking left matches if `_mode == JoinMode::Right` is not necessary.

mrzzzrm replied:
right!

mrzzzrm commented on src/lib/CMakeLists.txt:
```diff
@@ -206,6 +206,8 @@ set(
     operators/join_mpsm/radix_cluster_sort_numa.hpp
     operators/join_nested_loop.cpp
     operators/join_nested_loop.hpp
+    operators/join_verification_operator.cpp
```
suffix dropped. I thought it made it more clear that... this is an operator. But then, neither do "Sort", "GetTable", etc.

mrzzzrm commented:
> Any ideas why the coverage decreased?

Deleted some tests that the runner doesn't handle (e.g., for `description()`, `deep_copy()`). Added those again and also some more configurations. Coverage should be up slightly now.

mrzzzrm commented:
What does everyone think of the name "JoinTestRunner"? I don't like "Runner" - maybe just "Join[Operator]Test"?

mrks commented:
I‘m fine with `JoinTestRunner`. We should have something that shows that tests are generated, not hand-written. `AutomatedJoinTest`?

mrzzzrm commented:
"Automated" makes me think there is some intelligent logic generating the configurations. I'll think about it a bit and if nothing comes up I'll keep it as it is.

Bouncner commented:
Will take another look at it this evening. Please wait with the merge.

mrzzzrm commented:
> Please wait with the merge.

Will do, but if that's the case you can also just "Request Changes" to this PR with a pseudo-message which will automatically block merging.

Bouncner commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,590 @@
+      }
+    }
+
+    // The input tables are designed to have exclusive values. Test that these are handled correctly for different
```
What are exclusive values? Why does swapping (potentially) change the tested behavior?

mrzzzrm replied:
Added `// The input tables are designed to have exclusive values (i.e. values not contained in the other tables).`

Think of Semi/Anti joins. An exclusive value on the left side will not appear in the output of a Semi, but in the output of an Anti. Similarily, an exclusive value on the left side will have to be filled up with NULLs by a LeftOuter join, but discarded by a RightOuter join.

Bouncner commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,590 @@
+      }
+    }
+
+    // Test multi-predicate jin support for all join modes. Swap the input tables to trigger cases especially in the
```
"jin"

mrzzzrm replied:
gin

Bouncner commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,590 @@
+      }
+    }
+
+    // Materialization phases might take advantage of certain encodings. JSM, e.g., has a special path for Dictionaries
```
I think I've never seen the abbr. "JSM" before.

Bouncner replied:
finish with a dot.

mrzzzrm replied:
Strange, I use it all the time in my head. Changed to JoinSortMerge

Bouncner commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,590 @@
+      }
+    }
+
+    return configurations;
```
std::sort(std::unique(configurations))???

mrzzzrm replied:
You mean the beauty that is this?

```c++
    std::sort(configurations.begin(), configurations.end());
    configurations.erase(std::unique(configurations.begin(), configurations.end()), configurations.end());
```

But yeah, helps:

Before: 1115
After: 1003

Before: 827
After: 459

Before: 740
After: 682

Before: 653
After: 562

Before: 272
After: 233

Bouncner replied:
Beautiful. :)

mrzzzrm commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,590 @@
+      }
+    };
+
+    // JoinOperators (e.g. JoinHash) might pick a "common type"
```
Altered to 
```
    // JoinOperators (e.g. JoinHash) might pick a "common type" from the columns used by the primary predicate and cast
    // all values to that (e.g., joining float + long will cast everything to double)
```

mrzzzrm commented on src/test/operators/join_test_runner.cpp:
```diff
@@ -0,0 +1,590 @@
+
+    // TODO(anybody) include FrameOfReferenceEncoding once #1662 is resolved
+    const auto all_encoding_types =
+        std::vector{EncodingType::Unencoded, EncodingType::RunLength, EncodingType::Dictionary,
```
I know what you mean, but a "more noise!" approach ("use two encodings here for the default configuration") is snakeoil and shouldn't be necessary. We generate configurations that should trigger all paths and if we don't something is wrong with the configurations that we generate.

mrks replied:
Agreed
